internal/crypto/cert.go

```diff
@@ -37,32 +48,104 @@ func CNFromDeviceFingerprint(fingerprint string) (string, error) {
+	signCertificate(template *x509.Certificate, requestKey crypto.PublicKey) (*x509.Certificate, error)
+	EnsureClientCertificate(certFile, keyFile string, subjectName string, expireDays int) (*TLSCertificateConfig, bool, error)
+	MakeClientCertificate(certFile, keyFile string, subject string, expiryDays int) (*TLSCertificateConfig, error)
+	IssueRequestedClientCertificate(csr *x509.CertificateRequest, expirySeconds int) ([]byte, error)
+}
+
+
+
+var caMutex = &sync.Mutex{}
+var caList map[string]CA = make(map[string]CA)
+
+func EnsureCA(id string, cfg *config.Config) (CA, bool, error) {
+	caMutex.Lock()
+	defer caMutex.Unlock()
+
+	var cryptocfg *config.CryptographyConfig
+ 
+	
+	if cfg != nil {
+		cryptocfg = cfg.Cryptography
+	}
+
+	if cryptocfg == nil || len(cryptocfg.CAlist) == 0 {
+		ca_entry := &config.CryptographyConfigEntry{
+			CAType:config.InternalCA,
+			Id:DefaultCA,
+			InternalCAcfg: &config.InternalCAConfig{
+				Cert: CertFile(SignerCertName),
+				Key: KeyFile(SignerCertName),
+				Serial: "",
+				SignerName: SignerCertName,
+				ExpireDays: CaCertValidityDays,
+			},
+		}
+		cryptocfg = &config.CryptographyConfig{CAlist: []*config.CryptographyConfigEntry{ca_entry},}
+	}
 
-	SerialGenerator oscrypto.SerialGenerator
+	for _, data := range cryptocfg.CAlist {
+		configureCA(data)
+	}
+	return caList[id], false, nil
 }
 
-func EnsureCA(certFile, keyFile, serialFile, subjectName string, expireDays int) (*CA, bool, error) {
-	if ca, err := GetCA(certFile, keyFile, serialFile); err == nil {
-		return ca, false, err
+func configureCA(configData *config.CryptographyConfigEntry) error {
+	switch configData.CAType {
+		case config.InternalCA: {
+			if _, err := GetCA(configData.Id, configData.InternalCAcfg.Cert, configData.InternalCAcfg.Key, configData.InternalCAcfg.Serial); err == nil {
+				return err
+			}
+			_, err := MakeSelfSignedCA(configData.Id, configData.InternalCAcfg.Cert, configData.InternalCAcfg.Key, configData.InternalCAcfg.Serial, configData.InternalCAcfg.SignerName, int(configData.InternalCAcfg.ExpireDays))
+			return err
+		}
+		default:
+			return errors.New("Not Supported Yet")
 	}
-	ca, err := MakeSelfSignedCA(certFile, keyFile, serialFile, subjectName, expireDays)
-	return ca, true, err
+	return nil
 }
 
-func GetCA(certFile, keyFile, serialFile string) (*CA, error) {
+func GetCA(id, certFile, keyFile, serialFile string) (CA, error) {
+	var loaded CA
+
 	ca, err := oscrypto.GetCA(certFile, keyFile, serialFile)
 	if err != nil {
 		return nil, err
 	}
 	config := TLSCertificateConfig(*ca.Config)
-	return &CA{Config: &config, SerialGenerator: ca.SerialGenerator}, err
+	loaded = &internalCA{Config: &config, SerialGenerator: ca.SerialGenerator, Id:id,}
+	caList[id] = loaded
+	return loaded, err
 }
 
-func MakeSelfSignedCA(certFile, keyFile, serialFile, subjectName string, expiryDays int) (*CA, error) {
+func GetCachedCA(id string) (CA) {
+	caMutex.Lock()
+	defer caMutex.Unlock()
+	ca, ok := caList[id]
```

Comment:
No longer an issue. I switched to the k8 pattern of passing the CA/cfg top-down instead of having a singleton CA. I would not have coded it this way, but this seems to be the prevailing way of things in k8 land.